metomi/isodatetime/data.py

```diff
@@ -522,54 +542,50 @@ def __floordiv__(self, other):
                 "'%s' should be integer." %
                 type(other).__name__
             )
-        new = self.copy()
+        new = self._copy()
         if self.get_is_in_weeks():
-            new.weeks //= other
+            new._weeks //= other
             return new
-        new.years //= other
-        new.months //= other
-        new.days //= other
-        new.hours //= other
-        new.minutes //= other
-        new.seconds //= other
+        new._years //= other
+        new._months //= other
+        new._days //= other
+        new._hours //= other
+        new._minutes //= other
+        new._seconds //= other
         return new
 
+    def __hash__(self) -> int:
+        return hash(self.get_days_and_seconds())
+
     def __eq__(self, other: "Duration") -> bool:
-        my_data = self.get_days_and_seconds()
-        other_data = other.get_days_and_seconds()
-        return my_data == other_data
+        # TODO: check instance of other
+        return self.get_days_and_seconds() == other.get_days_and_seconds()
 
     def __ne__(self, other: "Duration") -> bool:
         return not self.__eq__(other)
 
     def __lt__(self, other: "Duration") -> bool:
-        my_data = self.get_days_and_seconds()
-        other_data = other.get_days_and_seconds()
-        return my_data < other_data
+        return self.get_days_and_seconds() < other.get_days_and_seconds()
 
     def __le__(self, other: "Duration") -> bool:
-        my_data = self.get_days_and_seconds()
-        other_data = other.get_days_and_seconds()
-        return my_data <= other_data
+        return self.get_days_and_seconds() <= other.get_days_and_seconds()
 
     def __gt__(self, other: "Duration") -> bool:
-        my_data = self.get_days_and_seconds()
-        other_data = other.get_days_and_seconds()
-        return my_data > other_data
+        return self.get_days_and_seconds() > other.get_days_and_seconds()
 
     def __ge__(self, other: "Duration") -> bool:
-        my_data = self.get_days_and_seconds()
-        other_data = other.get_days_and_seconds()
-        return my_data >= other_data
+        return self.get_days_and_seconds() >= other.get_days_and_seconds()
 
     def __bool__(self):
-        for attr in ["years", "months", "weeks", "days", "hours",
-                     "minutes", "seconds"]:
+        for attr in self.DATA_ATTRIBUTES:
             if getattr(self, attr, None):
                 return True
         return False
 
     def __str__(self):
+        if not self:
+            return "P0Y"
```

Comment:
Is this a bug fix?

Reply:
Just refactoring, previously it would go through a whole set of logic and only then to effectively check if any of the properties were not None